adhocracy4/comments_async/static/comments_async/comment_form.jsx

```diff
@@ -111,13 +112,13 @@ export default class CommentForm extends React.Component {
           </form>
         </div>
       )
-    } else if (!this.props.isReadOnly) {
+    } else if (!isAuthenticated && !this.props.isReadOnly) {
       return (
         <div className="a4-comments__login">
           <a href={config.getLoginUrl()}>{django.gettext('Please login to comment')}</a>
         </div>
       )
-    } else if (!this.props.isContextMember) {
+    } else if (isAuthenticated && !this.props.isContextMember) {
```

Comment:
Here we don't want isAuthenticated, because otherwise we would show the next option "The currently active state doesn't allow to comment" also to users that might be allowed to comment once they login. 
If we have a semi-private project for example and the currently active phase does allow invited participants to comment, we want to show this instead of the next option.

Reply:
think i am confused by the `isContextMember` flag, I thought it is depending on the user/member, but it can't be because its set even if no user is logged in.

Based on what information, this flag is set?

Reply:
You can find the definition here: https://github.com/liqd/adhocracy4/blob/fba25b7d8b2341de541592849ad0ba7295c25614/adhocracy4/comments_async/templatetags/react_comments_async.py#L74

The NormalUser is a dummy user defined here: https://github.com/liqd/adhocracy4/blob/master/adhocracy4/rules/discovery.py

So altogether isContextMember is true if the project is public (coming from the NormalUser part, because for public projects any authenticated user is context member) or the user is authenticated and is participant/moderator/org_member.

I hope this helped and wasnt even more confusing :grimacing: